specfile/changelog.py

```diff
@@ -0,0 +1,130 @@
+# Copyright Contributors to the Packit project.
+# SPDX-License-Identifier: MIT
+
+import collections
+import datetime
+import re
+from typing import List
+
+import arrow
+
+from specfile.exceptions import SpecfileException
+from specfile.sections import Section
+
+
+class ChangelogEntry:
+    """
+     Class that represents a changelog entry.
+
+    Attributes:
+        timestamp: Timestamp of the entry.
+        author: Author of the entry.
+        evr: EVR (epoch, version, release) of the entry.
+        content: List of lines forming the content of the entry.
+        legacy: Use legacy (date only) timestamp format.
+    """
+
+    def __init__(
+        self,
+        timestamp: datetime.datetime,
+        author: str,
+        evr: str,
+        content: List[str],
+        legacy: bool = False,
+    ) -> None:
+        self.timestamp = timestamp
+        self.author = author
+        self.evr = evr
+        self.content = content.copy()
+        self.legacy = legacy
+
+    def __str__(self) -> str:
+        return f"{self.header}\n" + "\n".join(self.content) + "\n"
+
+    def __repr__(self) -> str:
+        timestamp = repr(self.timestamp)
+        content = repr(self.content)
+        legacy = repr(self.legacy)
+        return f"ChangelogEntry({timestamp}, '{self.author}', '{self.evr}', {content}, {legacy})"
+
+    @property
+    def header(self) -> str:
+        """Entry header in the format that is being used in Fedora spec files."""
+        a = arrow.Arrow.fromdatetime(self.timestamp)
+        if not self.legacy:
+            dt = a.format("ddd MMM DD hh:mm:ss ZZZ YYYY")
+        else:
+            dt = a.format("ddd MMM DD YYYY")
+        return f"* {dt} {self.author} - {self.evr}"
+
+
+class Changelog(collections.UserList):
+    """
+    Class that represents a changelog.
+
+    Attributes:
+        data: List of individual entries.
+    """
+
+    def __str__(self) -> str:
+        return "\n".join(str(i) for i in reversed(self.data))
+
+    def __repr__(self) -> str:
+        data = repr(self.data)
+        return f"Changelog({data})"
+
+    @staticmethod
+    def parse(section: Section) -> "Changelog":
+        """
+        Parses a %changelog section.
+
+        Args:
+            section: Section to parse.
+
+        Returns:
+            Constructed instance of `Changelog` class.
+        """
+        header_regex = re.compile(
+            r"^\* \w+ (\w+ \d{2} (?:\d{2}:\d{2}:\d{2} [A-Z]+ )?\d{4}) (.+) - (.+)$"
```

Comment:
Nitpick: since this regex is getting a bit long (especially due to the date formats), it could be useful to use a verbose regex, would be a bit easier to understand. Or maybe at least add a comment with an example header line?

Reply:
I actually think using verbose regex would be less readable, I'll add example headers.

Reply:
It turns out there are other header formats in use, for example [kernel](https://src.fedoraproject.org/rpms/kernel/blob/rawhide/f/kernel.spec#_3005) has an unusual way to separate evr. I think we need to support this as well.